File: 18-Command/src/go/light_on_command.go
package main

// LightOnCommand binds "turn on" to a Light; Undo turns it off.
type LightOnCommand struct {
	light *Light
}

func NewLightOnCommand(light *Light) *LightOnCommand {
	return &LightOnCommand{light: light}
}

func (c *LightOnCommand) Execute() { c.light.On() }
func (c *LightOnCommand) Undo()    { c.light.Off() }

File: 18-Command/src/rust/light_on_command.rs
use std::rc::Rc;

use crate::command::Command;
use crate::light::Light;

/// Concrete Command — binds "turn on" to a Light; undo turns it off.
pub struct LightOnCommand {
    light: Rc<Light>,
}

impl LightOnCommand {
    pub fn new(light: Rc<Light>) -> Self {
        LightOnCommand { light }
    }
}

impl Command for LightOnCommand {
    fn execute(&self) {
        self.light.on();
    }
    fn undo(&self) {
        self.light.off();
    }
}
